File: src/libs/m3/arch/baremetal/VPE.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/CPU.h>
#include <base/ELF.h>
#include <base/Panic.h>

#include <m3/session/Pager.h>
#include <m3/stream/FStream.h>
#include <m3/vfs/MountTable.h>
#include <m3/vfs/RegularFile.h>
#include <m3/Syscalls.h>
#include <m3/VPE.h>

#include <stdlib.h>

namespace m3 {

void VPE::init_state() {
    if(Heap::is_on_heap(_eps))
        delete _eps;
    if(Heap::is_on_heap(_caps))
        delete _caps;

    _caps = reinterpret_cast<BitField<CAP_TOTAL>*>(env()->caps);
    if(_caps == nullptr)
        _caps = new BitField<CAP_TOTAL>();

    _eps = reinterpret_cast<BitField<EP_COUNT>*>(env()->eps);
    if(_eps == nullptr)
        _eps = new BitField<EP_COUNT>();

    _rbufcur = env()->rbufcur;
    _rbufend = env()->rbufend;
}

void VPE::init_fs() {
    if(Heap::is_on_heap(_fds))
        delete _fds;
    if(Heap::is_on_heap(_ms))
        delete _ms;

    if(env()->pager_sess && env()->pager_sgate)
        _pager = new Pager(env()->pager_sess, env()->pager_sgate, env()->pager_rgate);

    if(env()->mounts_len)
        _ms = MountTable::unserialize(reinterpret_cast<const void*>(env()->mounts), env()->mounts_len);
    else
        _ms = reinterpret_cast<MountTable*>(env()->mounts);

    if(env()->fds_len)
        _fds = FileTable::unserialize(reinterpret_cast<const void*>(env()->fds), env()->fds_len);
    else
        _fds = reinterpret_cast<FileTable*>(env()->fds);
}

Errors::Code VPE::run(void *lambda) {
    if(_pager)
        _pager->activate_rgate();

    Errors::Code err = copy_sections();
    if(err != Errors::NONE)
        return err;

    alignas(DTU_PKG_SIZE) Env senv;
    senv.pe = 0;
    senv.argc = env()->argc;
    senv.argv = RT_SPACE_START;
    senv.sp = CPU::get_sp();
    senv.entry = get_entry();
    senv.lambda = reinterpret_cast<uintptr_t>(lambda);
    senv.exitaddr = 0;

    senv.mounts_len = 0;
    senv.mounts = reinterpret_cast<uintptr_t>(_ms);
    senv.fds_len = 0;
    senv.fds = reinterpret_cast<uintptr_t>(_fds);
    senv.rbufcur = _rbufcur;
    senv.rbufend = _rbufend;
    senv.caps = reinterpret_cast<uintptr_t>(_caps);
    senv.eps = reinterpret_cast<uintptr_t>(_eps);
    senv.pager_sgate = 0;
    senv.pager_rgate = 0;
    senv.pager_sess = 0;

    senv._backend = env()->_backend;
    senv.pedesc = _pe;

    senv.heapsize = env()->heapsize;

    /* write start env to PE */
    _mem.write(&senv, sizeof(senv), RT_START);

    /* write args */
    char *buffer = static_cast<char*>(Heap::alloc(BUF_SIZE));
    size_t size = store_arguments(buffer, static_cast<int>(env()->argc),
        reinterpret_cast<const char**>(env()->argv));
    _mem.write(buffer, size, RT_SPACE_START);
    Heap::free(buffer);

    /* go! */
    return start();
}

Errors::Code VPE::exec(int argc, const char **argv) {
    alignas(DTU_PKG_SIZE) Env senv;
    char *buffer = static_cast<char*>(Heap::alloc(BUF_SIZE));

    if(_exec)
        delete _exec;
    _exec = new FStream(argv[0], FILE_RWX);
    if(!*_exec)
        return Errors::last;

    if(_pager)
        _pager->activate_rgate();

    uintptr_t entry;
    size_t size;
    Errors::Code err = load(argc, argv, &entry, buffer, &size);
    if(err != Errors::NONE) {
        Heap::free(buffer);
        return err;
    }

    senv.argc = static_cast<uint32_t>(argc);
    senv.argv = RT_SPACE_START;
    senv.sp = STACK_TOP;
    senv.entry = entry;
    senv.lambda = 0;
    senv.exitaddr = 0;

    /* add mounts, fds, caps and eps */
    /* align it because we cannot necessarily read e.g. integers from unaligned addresses */
    size_t offset = Math::round_up(size, sizeof(word_t));

    senv.mounts = RT_SPACE_START + offset;
    senv.mounts_len = _ms->serialize(buffer + offset, RT_SPACE_SIZE - offset);
    offset = Math::round_up(offset + static_cast<size_t>(senv.mounts_len), sizeof(word_t));

    senv.fds = RT_SPACE_START + offset;
    senv.fds_len = _fds->serialize(buffer + offset, RT_SPACE_SIZE - offset);
    offset = Math::round_up(offset + static_cast<size_t>(senv.fds_len), sizeof(word_t));

    size_t eps_caps = Math::round_up(sizeof(*_caps), sizeof(word_t)) +
        Math::round_up(sizeof(*_eps), sizeof(word_t));
    if(RT_SPACE_SIZE - offset < eps_caps)
        PANIC("State is too large");

    senv.caps = RT_SPACE_START + offset;
    memcpy(buffer + offset, _caps, sizeof(*_caps));
    offset = Math::round_up(offset + sizeof(*_caps), sizeof(word_t));

    senv.eps = RT_SPACE_START + offset;
    memcpy(buffer + offset, _eps, sizeof(*_eps));
    offset = Math::round_up(offset + sizeof(*_eps), DTU_PKG_SIZE);

    /* write entire runtime stuff */
    _mem.write(buffer, offset, RT_SPACE_START);

    Heap::free(buffer);

    senv.rbufcur = _rbufcur;
    senv.rbufend = _rbufend;

    /* set pager info */
    senv.pager_sess = _pager ? _pager->sel() : 0;
    senv.pager_sgate = _pager ? _pager->gate().sel() : 0;
    senv.pager_rgate = _pager ? _pager->rgate().sel() : 0;

    senv._backend = 0;
    senv.pedesc = _pe;

    senv.heapsize = _pager ? APP_HEAP_SIZE : 0;

    /* write start env to PE */
    _mem.write(&senv, sizeof(senv), RT_START);

    /* go! */
    return start();
}

void VPE::clear_mem(char *buffer, size_t count, uintptr_t dest) {
    memset(buffer, 0, BUF_SIZE);
    while(count > 0) {
        size_t amount = std::min(count, BUF_SIZE);
        _mem.write(buffer, Math::round_up(amount, DTU_PKG_SIZE), dest);
        count -= amount;
        dest += amount;
    }
}

Errors::Code VPE::load_segment(ElfPh &pheader, char *buffer) {
    if(_pager) {
        int prot = 0;
        if(pheader.p_flags & PF_R)
            prot |= Pager::READ;
        if(pheader.p_flags & PF_W)
            prot |= Pager::WRITE;
        if(pheader.p_flags & PF_X)
            prot |= Pager::EXEC;

        uintptr_t virt = pheader.p_vaddr;
        size_t sz = Math::round_up(pheader.p_memsz, static_cast<size_t>(PAGE_SIZE));
        if(pheader.p_memsz == pheader.p_filesz) {
            const RegularFile *rfile = static_cast<const RegularFile*>(_exec->file());
            return _pager->map_ds(&virt, sz, prot, 0, *rfile->fs(), rfile->fd(), pheader.p_offset);
        }

        assert(pheader.p_filesz == 0);
        return _pager->map_anon(&virt, sz, prot, 0);
    }

    /* seek to that offset and copy it to destination PE */
    size_t off = pheader.p_offset;
    if(_exec->seek(off, M3FS_SEEK_SET) != off)
        return Errors::INVALID_ELF;

    size_t count = pheader.p_filesz;
    size_t segoff = pheader.p_vaddr;
    while(count > 0) {
        size_t amount = std::min(count, BUF_SIZE);
        if(_exec->read(buffer, amount) != amount)
            return Errors::last;

        _mem.write(buffer, Math::round_up(amount, DTU_PKG_SIZE), segoff);
        count -= amount;
        segoff += amount;
    }

    /* zero the rest */
    clear_mem(buffer, pheader.p_memsz - pheader.p_filesz, segoff);
    return Errors::NONE;
}

Errors::Code VPE::load(int argc, const char **argv, uintptr_t *entry, char *buffer, size_t *size) {
    /* load and check ELF header */
    ElfEh header;
    if(_exec->read(&header, sizeof(header)) != sizeof(header))
        return Errors::INVALID_ELF;

    if(header.e_ident[0] != '\x7F' || header.e_ident[1] != 'E' || header.e_ident[2] != 'L' ||
        header.e_ident[3] != 'F')
        return Errors::INVALID_ELF;

    /* copy load segments to destination PE */
    uintptr_t end = 0;
    size_t off = header.e_phoff;
    for(uint i = 0; i < header.e_phnum; ++i, off += header.e_phentsize) {
        /* load program header */
        ElfPh pheader;
        if(_exec->seek(off, M3FS_SEEK_SET) != off)
            return Errors::INVALID_ELF;
        if(_exec->read(&pheader, sizeof(pheader)) != sizeof(pheader))
            return Errors::last;

        /* we're only interested in non-empty load segments */
        if(pheader.p_type != PT_LOAD || pheader.p_memsz == 0 || skip_section(&pheader))
            continue;

        load_segment(pheader, buffer);
        end = pheader.p_vaddr + pheader.p_memsz;
    }

    if(_pager) {
        // create area for stack and boot/runtime stuff
        uintptr_t virt = RT_START;
        Errors::Code err = _pager->map_anon(&virt, STACK_TOP - virt, Pager::READ | Pager::WRITE, 0);
        if(err != Errors::NONE)
            return err;

        // create heap
        virt = Math::round_up(end, static_cast<uintptr_t>(PAGE_SIZE));
        err = _pager->map_anon(&virt, APP_HEAP_SIZE, Pager::READ | Pager::WRITE, 0);
        if(err != Errors::NONE)
            return err;
    }

    *size = store_arguments(buffer, argc, argv);

    *entry = header.e_entry;
    return Errors::NONE;
}

size_t VPE::store_arguments(char *buffer, int argc, const char **argv) {
    /* copy arguments and arg pointers to buffer */
    uint64_t *argptr = reinterpret_cast<uint64_t*>(buffer);
    char *args = buffer + static_cast<size_t>(argc) * sizeof(uint64_t);
    for(int i = 0; i < argc; ++i) {
        size_t len = strlen(argv[i]);
        if(args + len >= buffer + BUF_SIZE)
            return Errors::INV_ARGS;
        strcpy(args, argv[i]);
        *argptr++ = RT_SPACE_START + static_cast<size_t>(args - buffer);
        args += len + 1;
    }
    return static_cast<size_t>(args - buffer);
}

}

File: src/libs/rust/src/vpe.rs
use alloc::boxed::FnBox;
use boxed::Box;
use cap::{CapFlags, Capability, Selector};
use cell::RefCell;
use cfg;
use com::{MemGate, RBufSpace};
use collections::Vec;
use core::{fmt, iter, intrinsics};
use dtu::{EP_COUNT, FIRST_FREE_EP, EpId};
use env;
use elf;
use errors::Error;
use heap;
use kif;
use kif::{CapType, CapRngDesc, INVALID_SEL, PEDesc};
use rc::Rc;
use session::M3FS;
use session::{Pager, Map};
use vfs::{BufReader, FileSystem, OpenFlags, Read, RegularFile, Seek, SeekMode};
use libc;
use syscalls;
use util;

pub struct VPE {
    cap: Capability,
    pe: PEDesc,
    mem: MemGate,
    next_sel: Selector,
    eps: u64,
    rbufs: RBufSpace,
    pager: Option<Pager>,
}

pub struct VPEArgs<'n, 'p> {
    name: &'n str,
    pager: Option<&'p str>,
    pe: PEDesc,
    muxable: bool,
}

pub trait Activity {
    fn vpe(&self) -> &VPE;

    fn start(&self) -> Result<(), Error> {
        syscalls::vpe_ctrl(self.vpe().sel(), kif::syscalls::VPEOp::Start, 0).map(|_| ())
    }

    fn stop(&self) -> Result<(), Error> {
        syscalls::vpe_ctrl(self.vpe().sel(), kif::syscalls::VPEOp::Stop, 0).map(|_| ())
    }

    fn wait(&self) -> Result<i32, Error> {
        syscalls::vpe_ctrl(self.vpe().sel(), kif::syscalls::VPEOp::Wait, 0)
    }
}

pub struct ClosureActivity<'v> {
    vpe: &'v mut VPE,
    _closure: env::Closure,
}

impl<'v> ClosureActivity<'v> {
    pub fn new<'vo : 'v>(vpe: &'vo mut VPE, closure: env::Closure) -> ClosureActivity<'v> {
        ClosureActivity {
            vpe: vpe,
            _closure: closure,
        }
    }
}

impl<'v> Activity for ClosureActivity<'v> {
    fn vpe(&self) -> &VPE {
        self.vpe
    }
}

impl<'v> Drop for ClosureActivity<'v> {
    fn drop(&mut self) {
        self.stop().ok();
        if let Some(ref mut pg) = self.vpe.pager {
            pg.deactivate();
        }
    }
}

pub struct ExecActivity<'v> {
    vpe: &'v mut VPE,
    _file: BufReader<RegularFile>,
}

impl<'v> ExecActivity<'v> {
    pub fn new<'vo : 'v>(vpe: &'vo mut VPE, file: BufReader<RegularFile>) -> ExecActivity<'v> {
        ExecActivity {
            vpe: vpe,
            _file: file,
        }
    }
}

impl<'v> Activity for ExecActivity<'v> {
    fn vpe(&self) -> &VPE {
        self.vpe
    }
}

impl<'v> Drop for ExecActivity<'v> {
    fn drop(&mut self) {
        self.stop().ok();
        if let Some(ref mut pg) = self.vpe.pager {
            pg.deactivate();
        }
    }
}

impl<'n, 'p> VPEArgs<'n, 'p> {
    pub fn new(name: &'n str) -> VPEArgs<'n, 'p> {
        VPEArgs {
            name: name,
            pager: None,
            pe: VPE::cur().pe(),
            muxable: false,
        }
    }

    pub fn pe(mut self, pe: PEDesc) -> Self {
        self.pe = pe;
        self
    }

    pub fn pager(mut self, pager: &'p str) -> Self {
        self.pager = Some(pager);
        self
    }

    pub fn muxable(mut self, muxable: bool) -> Self {
        self.muxable = muxable;
        self
    }
}

const VMA_RBUF_SIZE: usize  = 64;

static mut CUR: Option<VPE> = None;

impl VPE {
    fn new_cur() -> Self {
        let env = env::data();
        VPE {
            cap: Capability::new(0, CapFlags::KEEP_CAP),
            pe: env.pedesc.clone(),
            mem: MemGate::new_bind(1),
            // 0 and 1 are reserved for VPE cap and mem cap
            // it's initially 0. make sure it's at least the first usable selector
            next_sel: util::max(2, env.caps as Selector),
            eps: env.eps,
            rbufs: RBufSpace::new(env.rbuf_cur as usize, env.rbuf_end as usize),
            pager: match env.pager_sess {
                0 => None,
                s => Some(Pager::new_bind(s, env.pager_sgate, env.pager_rgate)),
            },
        }
    }

    pub fn cur() -> &'static mut VPE {
        unsafe {
            match CUR {
                Some(ref mut v) => v,
                None            => intrinsics::transmute(env::data().fds)
            }
        }
    }

    pub fn new(name: &str) -> Result<Self, Error> {
        Self::new_with(VPEArgs::new(name))
    }

    pub fn new_with(args: VPEArgs) -> Result<Self, Error> {
        let sels = VPE::cur().alloc_caps(2);

        let mut vpe = VPE {
            cap: Capability::new(sels + 0, CapFlags::empty()),
            pe: args.pe,
            mem: MemGate::new_bind(sels + 1),
            next_sel: 2,
            eps: 0,
            rbufs: RBufSpace::new(0, 0),
            pager: None,
        };

        let rbuf = if args.pe.has_mmu() {
            vpe.alloc_rbuf(util::next_log2(VMA_RBUF_SIZE) as usize)?
        }
        else {
            0
        };

        let pager = if args.pe.has_virtmem() {
            if let Some(p) = args.pager {
                Some(Pager::new(&mut vpe, rbuf, p)?)
            }
            else if let Some(p) = Self::cur().pager() {
                Some(p.new_clone(&mut vpe, rbuf)?)
            }
            else {
                None
            }
        }
        else {
            None
        };

        vpe.pager = if let Some(mut pg) = pager {
            let sgate_sel = pg.sgate().sel();
            let rgate_sel = match pg.rgate() {
                Some(rg) => rg.sel(),
                None     => INVALID_SEL,
            };

            // now create VPE, which implicitly obtains the gate cap from us
            vpe.pe = syscalls::create_vpe(
                vpe.sel(), vpe.mem().sel(), sgate_sel, rgate_sel, args.name,
                args.pe, vpe.alloc_ep()?, pg.rep(), args.muxable
            )?;

            // after the VPE creation, we can activate the receive gate
            // note that we do that here in case neither run nor exec is used
            pg.activate(vpe.sel())?;

            // mark the pager caps allocated
            vpe.next_sel = util::max(sgate_sel + 1, vpe.next_sel);
            if rgate_sel != INVALID_SEL {
                vpe.next_sel = util::max(rgate_sel + 1, vpe.next_sel);
            }
            // now delegate our VPE cap and memory cap to the pager
            pg.delegate_caps(&vpe)?;
            // and delegate the pager cap to the VPE
            vpe.delegate_obj(pg.sel())?;
            Some(pg)
        }
        else {
            vpe.pe = syscalls::create_vpe(
                sels + 0, sels + 1, INVALID_SEL, INVALID_SEL, args.name,
                args.pe, 0, 0, args.muxable
            )?;
            None
        };

        Ok(vpe)
    }

    pub fn sel(&self) -> Selector {
        self.cap.sel()
    }
    pub fn pe(&self) -> PEDesc {
        self.pe
    }
    pub fn mem(&self) -> &MemGate {
        &self.mem
    }

    pub fn pager(&self) -> Option<&Pager> {
        self.pager.as_ref()
    }

    pub fn alloc_cap(&mut self) -> Selector {
        self.alloc_caps(1)
    }
    pub fn alloc_caps(&mut self, count: u32) -> Selector {
        self.next_sel += count;
        self.next_sel - count
    }

    pub fn alloc_ep(&mut self) -> Result<EpId, Error> {
        for ep in FIRST_FREE_EP..EP_COUNT {
            if self.is_ep_free(ep) {
                self.eps |= 1 << ep;
                return Ok(ep)
            }
        }
        Err(Error::NoSpace)
    }

    pub fn is_ep_free(&self, ep: EpId) -> bool {
        ep >= FIRST_FREE_EP && (self.eps & (1 << ep)) == 0
    }

    pub fn free_ep(&mut self, ep: EpId) {
        self.eps &= !(1 << ep);
    }

    pub fn alloc_rbuf(&mut self, size: usize) -> Result<usize, Error> {
        self.rbufs.alloc(&self.pe, size)
    }
    pub fn free_rbuf(&mut self, addr: usize) {
        self.rbufs.free(addr)
    }

    pub fn delegate_obj(&mut self, sel: Selector) -> Result<(), Error> {
        self.delegate(CapRngDesc::new(CapType::OBJECT, sel, 1))
    }
    pub fn delegate(&mut self, crd: CapRngDesc) -> Result<(), Error> {
        let start = crd.start();
        self.delegate_to(crd, start)
    }
    pub fn delegate_to(&mut self, crd: CapRngDesc, dst: Selector) -> Result<(), Error> {
        syscalls::exchange(self.sel(), crd, dst, false)?;
        self.next_sel = util::max(self.next_sel, dst + crd.count());
        Ok(())
    }

    pub fn obtain(&mut self, crd: CapRngDesc) -> Result<(), Error> {
        let count = crd.count();
        let start = self.alloc_caps(count);
        self.obtain_to(crd, start)
    }

    pub fn obtain_to(&mut self, crd: CapRngDesc, dst: Selector) -> Result<(), Error> {
        let own = CapRngDesc::new(crd.cap_type(), dst, crd.count());
        syscalls::exchange(self.sel(), own, crd.start(), true)
    }

    pub fn revoke(&mut self, crd: CapRngDesc, del_only: bool) -> Result<(), Error> {
        syscalls::revoke(self.sel(), crd, !del_only)
    }

    pub fn run<F>(&mut self, func: Box<F>) -> Result<ClosureActivity, Error>
                  where F: FnBox() -> i32, F: Send + 'static {
        let get_sp = || {
            let res: usize;
            unsafe {
                asm!("mov %rsp, $0" : "=r"(res));
            }
            res
        };

        let sel = self.sel();
        if let Some(ref mut pg) = self.pager {
            pg.activate(sel)?;
        }

        let env = env::data();
        let mut senv = env::EnvData::default();

        senv.sp = get_sp() as u64;
        senv.entry = self.copy_regions(senv.sp as usize)? as u64;
        senv.lambda = 1;
        senv.exit_addr = 0;

        senv.heap_size = env.heap_size;

        senv.fds_len = 0;
        senv.fds = self as *const VPE as u64;

        // TODO
        // senv.mounts_len = 0;
        // senv.mounts = reinterpret_cast<uintptr_t>(_ms);
        // senv.fds_len = 0;
        // senv.fds = reinterpret_cast<uintptr_t>(_fds);
        // senv._backend = env()->_backend;

        // env goes first
        let mut off = cfg::RT_START + util::size_of_val(&senv);

        // create and write closure
        let closure = env::Closure::new(func);
        self.mem.write_obj(&closure, off)?;
        off += util::size_of_val(&closure);

        // write args
        senv.argc = env.argc;
        senv.argv = self.write_arguments(off, env::args())? as u64;

        // write start env to PE
        self.mem.write_obj(&senv, cfg::RT_START)?;

        // go!
        let act = ClosureActivity::new(self, closure);
        act.start().map(|_| act)
    }

    pub fn exec<S: AsRef<str>>(&mut self, fs: Rc<RefCell<M3FS>>,
                               args: &[S]) -> Result<ExecActivity, Error> {
        let file = fs.borrow_mut().open(args[0].as_ref(), OpenFlags::RX)?;
        let mut file = BufReader::new(file);

        let sel = self.sel();
        if let Some(ref mut pg) = self.pager {
            pg.activate(sel)?;
        }

        let mut senv = env::EnvData::default();

        senv.entry = self.load_program(&mut file)? as u64;
        senv.lambda = 0;
        senv.exit_addr = 0;
        senv.sp = cfg::STACK_TOP as u64;

        // write args
        let argoff = cfg::RT_START + util::size_of_val(&senv);
        senv.argc = args.len() as u32;
        senv.argv = self.write_arguments(argoff, args)? as u64;

        // TODO mounts, fds

        senv.rbuf_cur = self.rbufs.cur as u64;
        senv.rbuf_end = self.rbufs.end as u64;
        senv.caps = self.next_sel as u64;
        senv.eps = self.eps;
        senv.pedesc = self.pe();

        if let Some(ref pg) = self.pager {
            senv.pager_sess = pg.sel();
            senv.pager_sgate = pg.sgate().sel();
            senv.pager_rgate = match pg.rgate() {
                Some(rg) => rg.sel(),
                None     => INVALID_SEL,
            };
            senv.heap_size = cfg::APP_HEAP_SIZE as u64;
        }

        // write start env to PE
        self.mem.write_obj(&senv, cfg::RT_START)?;

        // go!
        let act = ExecActivity::new(self, file);
        act.start().map(|_| act)
    }

    fn copy_regions(&mut self, sp: usize) -> Result<usize, Error> {
        extern {
            static _text_start: u8;
            static _text_end: u8;
            static _data_start: u8;
            static _bss_end: u8;
            static heap_end: usize;
        }

        let addr = |sym: &u8| {
            (sym as *const u8) as usize
        };

        // use COW if both have a pager
        if let Some(ref pg) = self.pager {
            if Self::cur().pager().is_some() {
                return pg.clone().map(|_| unsafe { addr(&_text_start) })
            }
            // TODO handle that case
            unimplemented!();
        }

        unsafe {
            // copy text
            let text_start = addr(&_text_start);
            let text_end = addr(&_text_end);
            self.mem.write_bytes(&_text_start, text_end - text_start, text_start)?;

            // copy data and heap
            let data_start = addr(&_data_start);
            self.mem.write_bytes(&_data_start, libc::heap_used_end() - data_start, data_start)?;

            // copy end-area of heap
            let heap_area_size = util::size_of::<heap::HeapArea>();
            self.mem.write_bytes(heap_end as *const u8, heap_area_size, heap_end)?;

            // copy stack
            self.mem.write_bytes(sp as *const u8, cfg::STACK_TOP - sp, sp)?;

            Ok(text_start)
        }
    }

    fn clear_mem(&self, buf: &mut [u8], mut count: usize, mut dst: usize) -> Result<(), Error> {
        if count == 0 {
            return Ok(())
        }

        for i in 0..buf.len() {
            buf[i] = 0;
        }

        while count > 0 {
            let amount = util::min(count, buf.len());
            self.mem.write(&buf[0..amount], dst)?;
            count -= amount;
            dst += amount;
        }

        Ok(())
    }

    fn load_segment(&self, file: &mut BufReader<RegularFile>,
                    phdr: &elf::Phdr, buf: &mut [u8]) -> Result<(), Error> {
        file.seek(phdr.offset, SeekMode::SET)?;

        let mut count = phdr.filesz;
        let mut segoff = phdr.vaddr;
        while count > 0 {
            let amount = util::min(count, buf.len());
            let amount = file.read(&mut buf[0..amount])?;

            self.mem.write(&buf[0..amount], segoff)?;

            count -= amount;
            segoff += amount;
        }

        self.clear_mem(buf, phdr.memsz - phdr.filesz, segoff)
    }

    fn map_segment(&self, file: &mut BufReader<RegularFile>, pager: &Pager,
                   phdr: &elf::Phdr) -> Result<(), Error> {
        let mut prot = kif::Perm::empty();
        if (phdr.flags & elf::PF::R.bits()) != 0 {
            prot |= kif::Perm::R;
        }
        if (phdr.flags & elf::PF::W.bits()) != 0 {
            prot |= kif::Perm::W;
        }
        if (phdr.flags & elf::PF::X.bits()) != 0 {
            prot |= kif::Perm::X;
        }

        let size = util::round_up(phdr.memsz, cfg::PAGE_SIZE);
        if phdr.memsz == phdr.filesz {
            file.get_ref().map(pager, phdr.vaddr, phdr.offset, size, prot)
        }
        else {
            assert!(phdr.filesz == 0);
            pager.map_anon(phdr.vaddr, size, prot).map(|_| ())
        }
    }

    fn load_program(&self, file: &mut BufReader<RegularFile>) -> Result<usize, Error> {
        let mut buf = vec![0u8; 4096];
        let hdr: elf::Ehdr = file.read_object()?;

        if hdr.ident[0] != '\x7F' as u8 ||
           hdr.ident[1] != 'E' as u8 ||
           hdr.ident[2] != 'L' as u8 ||
           hdr.ident[3] != 'F' as u8 {
            return Err(Error::InvalidElf)
        }

        // copy load segments to destination PE
        let mut end = 0;
        let mut off = hdr.phoff;
        for _ in 0..hdr.phnum {
            // load program header
            file.seek(off, SeekMode::SET)?;
            let phdr: elf::Phdr = file.read_object()?;

            // we're only interested in non-empty load segments
            if phdr.ty != elf::PT::LOAD.val || phdr.memsz == 0 {
                continue;
            }

            if let Some(ref pg) = self.pager {
                self.map_segment(file, pg, &phdr)?;
            }
            else {
                self.load_segment(file, &phdr, &mut *buf)?;
            }
            off += hdr.phentsize as usize;

            end = phdr.vaddr + phdr.memsz;
        }

        if let Some(ref pg) = self.pager {
            // create area for boot/runtime stuff
            pg.map_anon(cfg::RT_START, cfg::RT_SIZE, kif::Perm::RW)?;

            // create area for stack
            pg.map_anon(cfg::STACK_BOTTOM, cfg::STACK_SIZE, kif::Perm::RW)?;

            // create heap
            pg.map_anon(util::round_up(end, cfg::PAGE_SIZE), cfg::APP_HEAP_SIZE, kif::Perm::RW)?;
        }

        Ok(hdr.entry)
    }

    fn write_arguments<I, S>(&self, off: usize, args: I) -> Result<usize, Error>
                             where I: iter::IntoIterator<Item = S>, S: AsRef<str> {
        let mut argptr = Vec::new();
        let mut argbuf = Vec::new();

        let mut argoff = off;
        for s in args {
            // push argv entry
            argptr.push(argoff);

            // push string
            let arg = s.as_ref().as_bytes();
            argbuf.extend_from_slice(arg);

            // 0-terminate it
            argbuf.push('\0' as u8);

            argoff += arg.len() + 1;
        }

        self.mem.write(&argbuf, off)?;
        self.mem.write(&argptr, argoff)?;
        Ok(argoff)
    }
}

impl fmt::Debug for VPE {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        write!(f, "VPE[sel: {}, pe: {:?}]", self.sel(), self.pe())
    }
}

pub fn init() {
    unsafe {
        CUR = Some(VPE::new_cur());
    }
}

pub fn reinit() {
    unsafe {
        CUR = None;
        VPE::cur().cap = Capability::new(0, CapFlags::KEEP_CAP);
        VPE::cur().mem = MemGate::new_bind(1);
    }
}
